src/main.cpp

```diff
@@ -862,16 +870,50 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
     }
 }
 
-bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidationState &state)
+bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidationState &state, uint32_t consensusBranchId)
 {
     // Basic checks that don't depend on any context
 
-    // Check transaction version
-    if (tx.nVersion < MIN_TX_VERSION) {
+    // Check min and max tx version of pre-Overwinter and Overwinter transactions.  Range allows for soft fork.
+    if (tx.fOverwintered) {
+        if (tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too low"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-low");
+        }
+        // We may want to have different rejection scores
+        else if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION ) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too high"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-high");
+        }
+    }
+    else if (tx.nVersion < MIN_TX_VERSION) {
```

Comment:
Hmm, it occurs to me that this change here is technically a change to the current consensus rules, because if this function is used anywhere on its own, it changes the allowed version range from `tx.header >= MIN_TX_VERSION` to `tx.header >= MIN_TX_VERSION || tx.header == -3` (using `tx.header` as a stand-in for what used to be internally named `tx.nVersion`. This means we need to be very careful to ensure we have somewhere a restriction `!fOverwinterActive && tx.header < MIN_TX_VERSION`.

Reply:
In `CheckTransactionWithoutProofVerification`